L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,81 @@
+
+* Name resolving. Built-in support is included for resolving from DNS, and a static resolver. Custom resolvers can be created.
+* Load balancing strategies. Built-in support is included for pick first and round robin load balancing. Custom load balancers can be created.
+* Connectivity status at sub-channel and channel level
+* Wait for ready
+
+## Proposal
+
+![Diagram](L80_graphics/diagram.png)
+
+### Initialization
+
+A channel is created with an address, an optional service provider, and an optional service config.
+
+```csharp
+var channel = GrpcChannel.ForAddress(
+    "dns://backend.default.svc.cluster.local",
+    new GrpcChannelOptions { ChannelCredentials = ChannelCredentials.Insecure });
+var client = new Greet.GreeterClient(channel);
+
+var response = await client.SayHelloAsync(new HelloRequest { Name = "world" });
+Console.WriteLine(response.Message);
+```
+
+* Address scheme will be used to create a resolver.
+  * Existing logic allows for `http` and `https`. They are special-cased to never use a resolver.
+  * A collection of resolver factories are fetched from the service provider. One is selected using the address scheme and it will create resolver. For example, `dns` will match to `DnsAddressResolverFactory` which will create a `DnsAddressResolver`.
+* Load balancer name in service configure will be used to create a load balancer.
+  * A collection of load balancer factories are fetched from the service provider. One is selected using the configured name (if none specified then `pick_first` is used) and it will create load balancer.
+  * Load balancer is initialized with an `IChannelController` instance by the channel. This is used to call back to the channel to create sub-channels and update the picker.
+
+### Connecting
+
+* No network activity until `GrpcChannel.ConnectAsync` is called, or the first RPC is started.
+* On connect, the `ConnectionManager` will subscribe to address resolver and call the resolver's `RefreshAsync` to prompt the resolver to run immediately.
+* Resolver will push addresses to `ConnectionManager` which will then call balancer.
+* Balancer will use `IChannelController` to create, remove or update sub-channels as needed based on the addresses.
+  * A new sub-channel will start connecting when `SubChannel.ConnectAsync` is called. A sub-channel will be shutdown when disposed.
+* Balancer will be notified on changes in sub-channel state. It will create a `SubChannelPicker` based on the current state.
+
+### Picking
+
+* `ConnectionManager.PickAsync` is called to get a ready sub-channel along with the sub-channel's connected address.
+* If a picker hasn't been set yet, or the picker or empty or an error then `PickAsync` will wait for a picker with a ready sub-channel. This depends on whether `WaitForReady` has been set.
```

Comment:
Added.

> Note that the picker should have the ability to explicitly tell the channel to queue the call, without necessarily returning an error

How do pickers in other implementations do this?

Reply:
The picker returns an enum indicating the type of the result, one of COMPLETE, QUEUE, FAIL, or DROP.  See my comment elsewhere about how to structure the pick result API.

Reply:
There are two differences between DROP and FAIL:
- If the picker returns FAIL, if the call has wait_for_ready=true, then the call actually gets queued instead of failing.  When the picker returns DROP, the call is failed regardless of whether or not it has wait_for_ready=true.
- If the picker returns FAIL and the call does has wait_for_ready=false, the call will be considered failed, but if retries are configured, it may be reattempted.  If the picker returns DROP, retries are inhibited.

DROP is used in xDS.  There are cases where the xDS server tells the client to drop a certain percentage of traffic, which should be honored regardless of the wait_for_ready value or retry configuration.